examples/advanced_operations/add_responsive_search_ad_full.py

```diff
@@ -0,0 +1,574 @@
+        customer_id=customer_id, operations=[*campaign_criterion_operations],
+    )
+    for results in ad_group_criterion_response.results:
+        print("Created keyword " f"{results.resource_name}.")
+
+    return
+
+
+def add_geo_targeting(client, customer_id, campaign_id):
+    """Creates geo targets.
+
+    Args:
+      client: an initialized GoogleAdsClient instance.
+      customer_id: a client customer ID.
+      campaign_id: an campaign ID.
+    
+    Returns:
+      Geo targets.
+    """
+    geo_target_constant_service = client.get_service("GeoTargetConstantService")
+    campaign_service = client.get_service("CampaignService")
+
+    # Search by location names from
+    # GeoTargetConstantService.suggest_geo_target_constants() and directly
+    # apply GeoTargetConstant.resource_name.
+    gtc_request = client.get_type("SuggestGeoTargetConstantsRequest")
+    gtc_request._LOCALE = _LOCALE
+    gtc_request._COUNTRY_CODE = _COUNTRY_CODE
+
+    # The location names to get suggested geo target constants.
+    gtc_request.location_names.names.extend(
+        [_GEO_LOCATION_1, _GEO_LOCATION_2, _GEO_LOCATION_3]
+    )
+
+    results = geo_target_constant_service.suggest_geo_target_constants(gtc_request)
+
+    operations = []
+    for suggestion in results.geo_target_constant_suggestions:
+        geo_target_constant = suggestion.geo_target_constant
+        print(
+            f"{geo_target_constant.resource_name} "
+            f"({geo_target_constant.name}, "
+            f"{geo_target_constant._COUNTRY_CODE}, "
+            f"{geo_target_constant.target_type}, "
+            f"{geo_target_constant.status.name}) "
+            f"is found in _LOCALE ({suggestion._LOCALE}) "
+            f"with reach ({suggestion.reach}) "
+            f"from search term ({suggestion.search_term})."
+        )
+        # Create the campaign criterion for location targeting.
+        campaign_criterion_operation = client.get_type("CampaignCriterionOperation")
+        campaign_criterion = campaign_criterion_operation.create
+        campaign_criterion.campaign = campaign_service.campaign_path(
+            customer_id, campaign_id
+        )
+        campaign_criterion.location.geo_target_constant = (
+            geo_target_constant.resource_name
+        )
+        operations.append(campaign_criterion_operation)
+
+    campaign_criterion_service = client.get_service("CampaignCriterionService")
+    campaign_criterion_response = campaign_criterion_service.mutate_campaign_criteria(
+        customer_id=customer_id, operations=[*operations]
+    )
+
+    for result in campaign_criterion_response.results:
+        print(f'Added campaign criterion "{result.resource_name}".')
+
+    return
+
+
+def add_images(client, customer_id, campaign_id):
+    # Step 6.1 - Add Image Asset.
+
+    # Download image from URL.
+    url = "https://img.cppng.com/download/2020-06/66912-logo-now-google-plus-search-free-transparent-image-hd.png"
```

Comment:
We have images hosted specifically for this purpose, for example: https://gaagl.page.link/bjYi

Reply:
TY!